### ISSUE TITLE
Memory leak raised by Sanitizer on GPU

### ISSUE THREAD

**Issue by guillaume-be** [LaurentMazare/tch-rs#300]
Hello,

Troubleshooting for https://github.com/guillaume-be/rust-bert/issues/112, I did run some experiments using the new Sanitizer and its memory leak detector (https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/sanitizer.html).

I could reproduce a small example which seems to identify a memory leak when variables are loaded on the GPU

```rust
use tch::{Device, Kind};
use tch::nn::VarStore;
use std::fs;

fn main() {
    let add = |vs: &tch::nn::Path| {
        let v = vs.sub("a").sub("b").ones("t2", &[3]);
        let u = vs.zeros("t1", &[4]);
        let _w = vs.sub("a").sub("b").sub("ccc").ones("t123", &[3]);
        let _w = vs.sub("a").sub("b").sub("ccc").ones("t123", &[3]);
        (u, v)
    };

    // No memory leak warning
    // let device = Device::Cpu;
    
    // Raises a memory leak warning
    let device = Device::cuda_if_available();
            
    let vs1 = VarStore::new(device);
    let (mut u1, mut v1) = add(&vs1.root());

}
```

Running `cargo run -Zbuild-std --target x86_64-unknown-linux-gnu path_to_this_binary` (with `export RUSTFLAGS=-Zsanitizer=leak`) indicates a leak of 20 bytes. Setting the device as Cpu does not raise this issue on the same machine (using CUDA 10.1 and tch-rs 0.3.0):

```
=================================================================
==108253==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x55dff5229910  (/home/tch-rs/target/x86_64-unknown-linux-gnu/debug/examples/test_memory_leak+0x31910)
    #1 0x7f43d322f2ce  (/home/libtorch/lib/libtorch_cuda.so+0x33312ce)

Direct leak of 8 byte(s) in 1 object(s) allocated from:
    #0 0x55dff5229910  (/home/tch-rs/target/x86_64-unknown-linux-gnu/debug/examples/test_memory_leak+0x31910)
    #1 0x7f43d322f2bd  (/home/libtorch/lib/libtorch_cuda.so+0x33312bd)

Indirect leak of 4 byte(s) in 1 object(s) allocated from:
    #0 0x55dff5229910  (/home/tch-rs/target/x86_64-unknown-linux-gnu/debug/examples/test_memory_leak+0x31910)
    #1 0x7f43d322f307  (/home/libtorch/lib/libtorch_cuda.so+0x3331307)

SUMMARY: LeakSanitizer: 20 byte(s) leaked in 3 allocation(s).
```

I am not convinced this is the source of the issue in https://github.com/guillaume-be/rust-bert/issues/112, but it may be worth having a look.

**Comment by LaurentMazare** [comment]
Hello,
Thanks for reporting this issue. It seems that you investigated this a bit more in https://github.com/guillaume-be/rust-bert/issues/112 and now think that some of these could be some spurious warnings. Does it also apply here?

**Comment by guillaume-be** [comment]
Hello,

This may very well be a spurious error. I have made a few additional checks with valgrind, and the same memory leak is identified when using CUDA, but not when using the CPU. My guess is that these memory profilers lose track of the memory the moment it moves to the GPU buffer. Running the code in a loop does not lead to increase memory consumption (including the GPU buffer), so I am assuming this is not an issue.

To reproduce, I used the following script to generate two valgrind logs: one on CPU and one on CUDA:
```rust
use tch::{Device, nn};

fn main() {
    let device = Device::Cpu;
    let vs = nn::VarStore::new(device);
    let _module = nn::linear(&vs.root() / "dense", 1024, 1024, Default::default());
}
```
- [CUDA logs](https://gist.github.com/guillaume-be/7119c1d13c0409dd7b3e6fda48ab4c51)
- [CPU logs](https://gist.github.com/guillaume-be/053dce6404c204431152ece8c7be6ca4)

I raised the issue as part of the investigations in https://github.com/guillaume-be/rust-bert/issues/112 since the only potential source of memory leak seemed to be at the FFI level. We could not confirm a memory leak based on a number of experiments, so my assumption is that the warnings raised by the LeakSanitizer and Valgrind on GPU are both spurious. 

Thank you for following up on this - I suggest closing this issue for now unless you believe this is something worth looking at more into details.
